---
title: "R/qtl2pattern Vignette"
author: "Brian S. Yandell"
date: "2/8/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This continues and example developed in library `qtl2ggplot` using tools found in `qtl2pattern`.

```{r}
library(qtl2)
library(qtl2ggplot)
library(qtl2pattern)
library(ggplot2)
```

```{r}
if(!file.exists("DOex.zip")) {
  dirpath <- paste0("https://raw.githubusercontent.com/rqtl/",
                    "qtl2data/master/DOex")
}
DOex <- read_cross2("DOex.zip")
```

Create artificial second phenotype as arcsic sqrt of first one.

```{r}
DOex$pheno <- cbind(DOex$pheno, asin = asin(sqrt(DOex$pheno[,1]/100)))
DOex$pheno[,"asin"] <- DOex$pheno[,"asin"] *
  sd(DOex$pheno[,"OF_immobile_pct"], na.rm = TRUE) /
  sd(DOex$pheno[,"asin"], na.rm = TRUE)
```

Subset to chr 2.

```{r}
DOex <- DOex[,"2"]
```

Calculate genotype probabilities and convert to allele probabilities

```{r}
pr <- calc_genoprob(DOex, error_prob=0.002)
apr <- genoprob_to_alleleprob(pr)
```

Download snp info from web

```{r}
if(!file.exists(filename <- "c2_snpinfo.rds")) {
  filename <- file.path("https://raw.githubusercontent.com/rqtl",
                        "qtl2data/master/DOex", 
                        "c2_snpinfo.rds")
  tmpfile <- tempfile()
  download.file(filename, tmpfile, quiet=TRUE)
  snpinfo <- readRDS(tmpfile)
  unlink(tmpfile)
} else {
  snpinfo <- readRDS(filename)
}
snpinfo <- 
  dplyr::rename(
    snpinfo,
    pos = pos_Mbp)
```

Calculate strain distribution patterns

```{r}
snpinfo$sdp <- calc_sdp(snpinfo[,-(1:4)])
```

Create index to non-equivalent set of SNPs.

```{r}
snpinfo <- index_snps(DOex$pmap, snpinfo)
```

Convert to snp probabilities

```{r}
snpinfo <- index_snps(DOex$pmap, snpinfo)
snppr <- genoprob_to_snpprob(apr, snpinfo)
```

## Genome Scan

Package `qtl2pattern` has `summary` methods for `scan1` and `scan1coef`.
Not sure these are really needed. The `R/qtl2` routine `find_peaks` takes care of some but not all of this functionality.

```{r}
scan_apr <- scan1(apr, DOex$pheno)
```

```{r}
find_peaks(scan_apr, DOex$pmap)
```

```{r}
summary(scan_apr, DOex$pmap)
```

```{r}
coefs <- scan1coef(apr, DOex$pheno)
```

```{r}
summary(coefs, scan_apr, DOex$pmap)
```

## SNP Association Mapping

Perform SNP association analysis (here, ignoring residual kinship)

```{r}
scan_snppr <- scan1(snppr, DOex$pheno)
```

```{r}
summary(scan_snppr, DOex$pmap, snpinfo)
```

## Scan Patterns

One relatively new and somewhat confusing addition is **scan patterns**, in which the contrast pattern for top SNPs is propagated over the interval to "fill in", leading to LOD and contrast scans. That is, only a few SNPs may show a particular pattern, but key differences might be seen nearby if we could impute SNPs of the same pattern.

```{r}
top_snps_tbl <- top_snps_all(scan_snppr, snpinfo)
```

```{r}
(patterns <- dplyr::arrange(
  summary(top_snps_tbl),
  dplyr::desc(max_lod)))
```

```{r}
(pheno_name <- dimnames(DOex$pheno)[[2]][1])
```

```{r}
scan_pat <- scan_pattern(pr, DOex$pheno,
                         map = DOex$pmap,
                         patterns = patterns)
```

**This table should not have missing values.**

```{r}
summary(scan_pat, DOex$pmap)
```

Modify plot_scan_pattern to facet on multiple phenotypes.

```{r}
autoplot(scan_pat, DOex$pmap) +
  geom_vline(xintercept = c(96.5,98.5))
```

```{r}
autoplot(scan_pat, DOex$pmap, "coef", ylim=c(-100,100)) +
  geom_vline(xintercept = c(96.5,98.5))
```

```{r}
autoplot(scan_pat, DOex$pmap, "coef_and_lod", ylim=c(-100,100))
```

Here we show the traditional QTL effects and LOD scan next to the newer scan pattern effects and LOD scans for comparison. The traditional LOD has 7 model degrees of freedom (df), while the scan pattern contrast LODs each have 2 model df. The LOD curves drop off somewhat more steeply for the scan pattern contrasts, and the peak is almost as high (`r round(max(scan_pat$scan, DOex$pmap)[,3], 1)` vs. `r round(max(scan_apr, DOex$pmap)[,3], 1)`).

The traditional QTL effects compare 8 founder additive allele effects,
while each scan pattern contrast compares the 3 SNP alleles,
reference (`ref` = `dark green`),
heterozygote (`het` = `dark orange`) and
alternate (`alt` = `dark blue`).
The scan pattern contrasts appear at first to be confusing and contradictory, with the `ABCDGH:EF` contrast showing that the `ref` allele is dominant (with `het` at the same level), while the `ABCDFGH:E` suggests `ref` is recessive. However, a closer look shows the following story with regard to the `alt` allele:

* the `ABCDFGH:E` contrast finds 1 or 2 `E=NZO` alleles have lower response
* the `ABCDGH:EF` contrast finds having only `E=NZO` and/or `F=CAST` alleles leads to lower response

These findings are thus complementary, and they agree with the traditional QTL effects plot showing `NZO` = `powder blue` and `CAST` = `green` alleles being much lower than the other 6 alleles.

```{r}
autoplot(coefs, DOex$pmap, scan1_output = scan_apr)
```

```{r}
autoplot(scan_pat, DOex$pmap, "coef_and_lod", ylim = c(-100,100))
```
